### PR TITLE
[back] perf: reduce nbr of SQL queries in _get_user_comparability_augmenting_videos

### PR DESCRIPTION
**Pull request by lfaucon**
I believe this does the intended and reduces the number of request. This happens at every call to the suggestion provider.

### CODE REVIEW

File: backend/tournesol/suggestions/suggestionprovider.py [hunk 1]
```diff
@@ -61,6 +61,9 @@ def __init__(self, actual_poll: Poll):
 
         # create required user graphs (none at first in fact)
 
+    # TODO: This method is expensive (it fetches > 10k comparisons). It should probably be cached
+    # since it is used in every call of SuggestionProvider.get_first_video_recommendation and
+    # SuggestionProvider.get_second_video_recommendation
     def _get_user_comparability_augmenting_videos(self) -> list[SuggestedVideo]:
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/tournesol/suggestions/suggestionprovider.py:67]: Just for information, I checked the code:

```python
supertrusted_comparisons = Comparison.objects.filter(
    poll=self.poll,
    user__in=User.supertrusted_seed_users()
).values_list("entity_1__uid", "entity_2__uid")
```
... and it has the following performance (looks OK to me):

```
 Nested Loop  (cost=1089.86..1231.78 rows=14 width=28) (actual time=6.550..89.764 rows=11474 loops=1)
   ->  Nested Loop  (cost=1089.57..1221.90 rows=14 width=22) (actual time=6.532..56.083 rows=11474 loops=1)
         ->  Nested Loop  (cost=1089.29..1212.02 rows=14 width=16) (actual time=6.514..19.858 rows=11474 loops=1)
               ->  HashAggregate  (cost=1088.87..1088.90 rows=3 width=8) (actual time=6.460..6.484 rows=6 loops=1)
                     Group Key: v0.id
                     Batches: 1  Memory Usage: 24kB
                     ->  Hash Join  (cost=33.40..1088.86 rows=3 width=8) (actual time=0.970..6.448 rows=6 loops=1)
                           Hash Cond: (lower(regexp_replace((v0.email)::text, '(.*)(@.*$)'::text, '\2'::text)) = (u0.domain)::text)
                           ->  Seq Scan on core_user v0  (cost=0.00..1055.44 rows=7 width=31) (actual time=0.042..5.378 rows=7 loops=1)
                                 Filter: is_staff
                                 Rows Removed by Filter: 12137
                           ->  Hash  (cost=26.81..26.81 rows=527 width=13) (actual time=0.690..0.692 rows=543 loops=1)
                                 Buckets: 1024  Batches: 1  Memory Usage: 33kB
                                 ->  Seq Scan on core_emaildomain u0  (cost=0.00..26.81 rows=527 width=13) (actual time=0.030..0.498 rows=543 loops=1)
                                       Filter: ((status)::text = 'ACK'::text)
                                       Rows Removed by Filter: 810
               ->  Index Scan using tournesol_comparison_user_id_poll_id_video_1__94c4dd36_uniq on tournesol_comparison  (cost=0.42..40.94 rows=10 width=24) (actual time=0.028..1.695 rows=1912 loops=6)
                     Index Cond: ((user_id = v0.id) AND (poll_id = 1))
         ->  Index Scan using tournesol_video_pkey on tournesol_entity  (cost=0.29..0.71 rows=1 width=22) (actual time=0.003..0.003 rows=1 loops=11474)
               Index Cond: (id = tournesol_comparison.entity_1_id)
   ->  Index Scan using tournesol_video_pkey on tournesol_entity t5  (cost=0.29..0.71 rows=1 width=22) (actual time=0.002..0.002 rows=1 loops=11474)
         Index Cond: (id = tournesol_comparison.entity_2_id)
 Planning Time: 1.853 ms
 Execution Time: 90.971 ms
```

See the complete SQL query

```sql
SELECT 
  "tournesol_entity"."uid", 
  T5."uid" 
FROM 
  "tournesol_comparison" 
  INNER JOIN "tournesol_entity" ON (
    "tournesol_comparison"."entity_1_id" = "tournesol_entity"."id"
  ) 
  INNER JOIN "tournesol_entity" T5 ON (
    "tournesol_comparison"."entity_2_id" = T5."id"
  ) 
WHERE 
  (
    "tournesol_comparison"."poll_id" = 1 
    AND "tournesol_comparison"."user_id" IN (
      SELECT 
        V0."id" 
      FROM 
        "core_user" V0 
      WHERE 
        (
          EXISTS(
            SELECT 
              (1) AS "a" 
            FROM 
              "core_emaildomain" U0 
            WHERE 
              (
                U0."domain" = (
                  LOWER(
                    regexp_replace(
                      V0."email", 
                      '(.*)(@.*$)', 
                     '\2'
                    )
                  )
                ) 
                AND U0."status" = 'ACK'
              ) 
            LIMIT 
              1
          ) AND V0."is_staff"
        )
    )
  );
```

**Comment by GresilleSiffle** [comment]
Thank you @lfaucon , this is a welcome optimization.

The SQL query doesn't seem to be slow, and I'm not sure that returning a list of 10k elements is problematic here in term of performance. I didn't investigate deeply the consequences of using such a list, but when calling the endpoint `entities_to_compare` the "bottleneck" seems to be before the call to `_get_user_comparability_augmenting_videos()`.

File: backend/tournesol/suggestions/suggestionprovider.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (amatissart) [reply backend/tournesol/suggestions/suggestionprovider.py:67]: The version on "main" uses 2 `LEFT OUTER JOIN` on "tounesol_comparison", which is very suboptimal (even if we had alreay optimized it in #989 )

The performance gain seems to be significant on my local dev-env

* `master`
```
In [4]: %time videos = sp._get_user_comparability_augmenting_videos()
CPU times: user 5.55 ms, sys: 751 µs, total: 6.3 ms
Wall time: 13.6 s
```

* This branch (02e23b74ab)
```
In [33]: %time videos = sp._get_user_comparability_augmenting_videos()
CPU times: user 7.86 ms, sys: 1 ms, total: 8.86 ms
Wall time: 29.2 ms
```

So it seems to be a very good catch @lfaucon :+1: 
And a sad reminder that Django ORM can behave very badly, even on relatively simple queries.

**Review by amatissart — APPROVED** [review]
As dicussed on Wednesday, this is a good first step and can be merged as is, before addressing https://github.com/tournesol-app/tournesol/issues/1244

**Pull request merged**